### PR TITLE
Rewrite indentation code

### LINKED ISSUE #1: Add unit tests
Unit tests could help especially with the indent fn. There are probably a few cases broken currently: https://github.com/guns/vim-clojure-static/pull/83

Not sure how these are usually done with Vimscript.

We could probably copy test cases from some other project for a start:
https://github.com/clojure-emacs/clojure-mode/blob/master/test/clojure-mode-indentation-test.el
https://github.com/weavejester/cljfmt/blob/master/cljfmt/test/cljfmt/core_test.cljc#L11

### LINKED ISSUE #6: Improve indent performance
https://github.com/guns/vim-clojure-static/issues/88

With 6b18b4e4ad1c14dd22c9a33974e16e1bd3eba2d7 test script, master: 

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
29646              4.656598  <SNR>35_syn_id_name()
 1532   5.261001   0.407066  <SNR>35_match_pairs()
29028   4.712738   0.186393  <SNR>35_ignored_region()
28332   4.856748   0.155025  <SNR>35_is_paren()
30068              0.059254  <SNR>35_current_char()
  417   5.570994   0.033857  GetClojureIndent()
  417   5.439542   0.032423  <SNR>35_clojure_indent_pos()
  315   1.959061   0.012072  <SNR>35_clojure_is_method_special_case_worker()
  416   0.112460   0.006156  <SNR>35_clojure_check_for_string_worker()
  416   0.116510   0.004050  <SNR>35_check_for_string()
  581              0.003795  <SNR>35_strip_namespace_and_macro_chars()
  130              0.003790  <SNR>35_match_one()
  581              0.003502  <SNR>35_current_word()
  315   1.962500   0.003439  <SNR>35_is_method_special_case()
 1395              0.002532  <SNR>35_bracket_type()
  315              0.001042  <SNR>35_is_reader_conditional_special_case()
```

Because indent uses the syntax rules, #5 branch already improves this a quite bit:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
29646              1.478719  <SNR>35_syn_id_name()
 1532   2.199816   0.404030  <SNR>35_match_pairs()
29028   1.617763   0.185423  <SNR>35_ignored_region()
28332   1.798798   0.156209  <SNR>35_is_paren()
30068              0.059659  <SNR>35_current_char()
  417   2.391534   0.034190  GetClojureIndent()
  417   2.312287   0.032532  <SNR>35_clojure_indent_pos()
  315   0.809202   0.011783  <SNR>35_clojure_is_method_special_case_worker()
  416   0.046043   0.006287  <SNR>35_clojure_check_for_string_worker()
  130              0.004105  <SNR>35_match_one()
  581              0.004033  <SNR>35_strip_namespace_and_macro_chars()
  416   0.050047   0.004003  <SNR>35_check_for_string()
  581              0.003524  <SNR>35_current_word()
  315   0.812542   0.003340  <SNR>35_is_method_special_case()
 1395              0.002603  <SNR>35_bracket_type()
  315              0.001093  <SNR>35_is_reader_conditional_special_case()
```

### LINKED ISSUE #21: Question: is there a way to configure Tonsky's indentation rule?
Context: https://tonsky.me/blog/clojurefmt/

### LINKED ISSUE #33: Compatibility with tree-sitter
Hey friends,

The indentation system is built on the syntax highlighting to know how to interpret things like `\{`. If I use [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) (which relies on [this grammar](https://github.com/sogaiu/tree-sitter-clojure/)) and disable nvim's syntax highlighting, indentation is completely broken.

I don't want to lose the nice indentation provided by this plugin (as there is no indentation information for Clojure in nvim-treesitter), but enabling both vimscript syntax highlighting and treesitter highlighting is bad and slow.

Any suggestions on how to handle this?

### LINKED ISSUE #34: [Bug] incorrect indentation for curly braces with `=`
When typing, indentation is correct, but when indenting using something like `gg=G`, the indentation becomes incorrect. The following example is in Fennel, but the idea is the same.

Correct indentation:
```clojure
{1 :gbprod/cutlass.nvim
 :opts {:cut_key :x
        :exclude [:ns :nS]
        :registers {:change :c :delete :d :select :s}}}
```
After `gg=G`:
```clojure
{1 :gbprod/cutlass.nvim
:opts {:cut_key :x
:exclude [:ns :nS]
:registers {:change :c :delete :d :select :s}}}
```

### LINKED ISSUE #39: Run indentation tests on Neovim too
With the new indentation test mechanism (#26) it should be possible to run the indentation tests in Neovim too.

### LINKED ISSUE #40: [Feature Request] Set `lisp` option
Some plugins use `set lisp?` to determine if the file is Lisp or not. It would be cool if you had `lisp` option set just so that outside plugins may detect Clojure as Lisp. It's possible for outside plugins to hardcode Clojure as Lisp, but I don't think that's the best approach.

### PR DESCRIPTION
**Pull request by axvr**
> [!Tip]
> As of the latest change, indentation is 2–3x faster with the potential to surpass 10x with Vim9script and Lua.

This PR contains a rewrite of the entire Clojure indentation code that aims to:

1. Improve the (currently abysmal) indent performance and unlock yet more improvements in future,
   - #6
   - #32
   - #35
2. Change the defaults to be more similar to the [Clojure Style Guide](https://guide.clojure.style), [cljfmt](https://github.com/weavejester/cljfmt), [Emacs](https://github.com/clojure-emacs/clojure-mode) and [VS Code](https://calva.io/) defaults,
   - #22
3. Enable greater configuration of indent rules (such that various indentation styles are possible),
   - #21
5. Simplify the indentation code as much as possible (currently so complex, it is near impossible to work on),
6. Improve accuracy of the indentation algorithms, eliminating most (if not all) previous indentation bugs,
   - #34
   - https://github.com/clojure-vim/clojure.vim/pull/31#issuecomment-2227410501
7. Eliminate reliance on syntax highlighting to be compatible with alternate syntax highlighting systems (e.g. Tree-sitter),
   - #33
9. Create a rich indentation test suite,
   - #1
   - #39
   - #26
10. Make quality of life improvements (when the (Neo)Vim version supports them), such as preventing `=` operator from altering the indentation of multi-line strings and setting the `'lisp'` option in Clojure buffers.
    - #40

## To do

- [x] New indentation algorithm.
- [x] More indentation tests + test the different config options.
- [x] Run indentation tests on Vim and Neovim.
- [x] Update documentation (doc file and README).
- [x] Replace existing configuration options.
- [ ] Resolve most remaining TODOs.
- [ ] Create a GitHub issue for the remaining missing indent features:
  - Indentation rules for reader conditionals (i.e. `#?@(...)`, `#?(...)`).
  - Complex indentation rules (e.g. `letfn`, `extend-protocol`, etc.).
- [ ] Maybe write blog post about this update?

### CODE REVIEW

**Comment by axvr** [comment]
I've been experimenting with a custom algorithm as per the check list item above.  This version eliminates all syntax highlight checks (and doesn't require any hacks to make those checks work) plus it is both faster and less code still.

Another benefit of this one is that since the bottle neck is no longer syntax highlight checks, by writing a version of the algorithm in Vim9 script, we can get it even faster still!  (Maintaining both a legacy Vim script and Vim9 script versions should be feasible due to how much simpler this is than the original implementation.)

TL;DR: I'll push a new version to this branch in a few days which will be:
1. Faster,
2. Have potential to become significantly faster (Vim9 script),
3. Simpler with no hacks,
4. Less code,
5. Works even when syntax is off.

Once this branch is feature complete, I'll share benchmarks comparing it to the original as is on `master`.

**Comment by axvr** [comment]
Performance benchmarks as of current change.  This is measuring the time to reindent a whole file on Macbook Pro 2021, so expect much more noticable performance gains on slower hardware.

| Code version | 200 lines of Clojure | 600 lines of EDN |
|---|---|---|
| Original | 0.96s | 3.5s |
| New (legacy) | 0.33s | 2.6s |
| New (Vim9) | 0.06s | 0.46s |

(I'll share new benchmarks (and the full reports generated by Vim) once this branch is feature complete.)

With the Vim9 implementation in the code too (not pushed yet, still figuring out how best to integrate it in a maintainable way), we can expect the file size of `indent/clojure.vim` to be ~150 LOC less than the original.

**Comment by NoahTheDuke** [comment]
Any updates on this?

**Comment by axvr** [comment]
@NoahTheDuke yes, slow steady progress during the limited time I have to work on it at the moment.  GitHub doesn't show it well, but there have been a bunch of bug fixes over the last few weeks.  Currently my focus in on getting the majority of the tests to pass.

If you wanted to use it, this branch is actually usable if you don't mind a few minor formatting issues.

**Comment by NoahTheDuke** [comment]
Cool. I mistakenly trusted Github's "added 11 commits 2 months ago", which it seems bundles recent stuff too. I don't mean to pressure, just saw a bunch of activity then looked like nothing. Glad to hear you're still poking at it. I'll give it a whirl locally!

**Comment by NoahTheDuke** [comment]
Just tried it out and it's great so far. Thanks so much!

**Comment by axvr** [comment]
Update: nearing readiness!  :tada:

This branch has now surpassed the original indentation code in pretty much every aspect (~2–3x faster, less and simpler code, more accurate indentation, better defaults, more customisable, not dependent on syntax highlighting, etc.).  There are a few small known bugs left to fix, the documentation needs updating and a little bit of code clean up will be required.

(The greatly expanded test suite is also already [paying dividends](https://idioms.thefreedictionary.com/pay+dividends); helping catch accidental regressions!  More tests to come soon.)

**Comment by SavageMessiah** [comment]
Thanks for working on this, I've been trying clojure dev in neovim lately and was getting really frustrated by the fact that it would indent based on brackets in string literals. I just tried this branch and that problem is fixed.

**Comment by axvr** [comment]
> Thanks for working on this, I've been trying clojure dev in neovim lately and was getting really frustrated by the fact that it would indent based on brackets in string literals. I just tried this branch and that problem is fixed.

@SavageMessiah great to hear that!  If you encounter any issues on this branch please let me know.

**Comment by NoahTheDuke** [comment]
Feedback:

```clojure
(filter-test
 a b c)

(filter-suite
  a b c)

(letfn [(filter-test
         [test-case]
         ...)
        (filter-suite
          [suite]
          ...)]
  ...)
```

It seems the matching algorithm for function indentation is based on any portion of the function's name, not the whole thing, so `test` is indented with 1 space and others are indented with 2. Regardless of my preference (1 vs 2), these should not be indented differently, especially when they're both user-defined functions.

Additionally, `letfn` functions have hanging indentation, not flow indentation, as seen below. The `func` body should be indented 2 spaces, not even with the brackets.

```clojure
(letfn [(func [test-case]
              ...)]
  ...)
```

Thanks so much!

**Comment by axvr** [comment]
Hi @NoahTheDuke, thanks for raising and testing!

> It seems the matching algorithm for function indentation is based on any portion of the function's name, not the whole thing, so test is indented with 1 space and others are indented with 2. [...]

Unfortunately I'm not seeing the same behaviour you are with `filter-test` vs. `filter-suite`, perhaps you have some config options set controlling the indentation for those?  If you do and can share it, I'll happily take a further look into this issue.

> Additionally, letfn functions incorrectly have hanging indentation, not flow indentation [...]

I am aware of this current deficiency in the indentation algorithm.  It is possible to solve, but it will be tricky as `letfn` falls into what I call _complex indentation_, which includes others such as `extend-protocol`, `extend-type`, and method bodies of `defrecord`, `deftype` and `defprotocol`.  I call it "complex indentation" as they alter the indentation rules of the lists within them.

I will fix it, but it will likely be **after** this PR is merged as their use isn't particularly common and hopefully the other improvements I've made will make up for it being a temporary issue.  As a partial workaround, I'd recommend using `let` and `fn` instead of `letfn` _where possible_, as it's often no more code and is (at least in my opinion) more readable too.  Alternatively you can run `cljfmt` before committing or hook it into Vim with a pre-write autocommand.

---

### Asides

<details>
<summary>
Why is indenting <code>letfn</code> so difficult?  (Click here for more info.)
</summary>
<br>
<p>
The challenge with the complex indentation rules is in making it both efficient within Vim script, and generic/customisable.
</p>
<p>
Emacs' clojure-mode avoids the first (by Emacs Lisp being much, much faster than Vim script :sob:) and solves the latter nicely with a way of defining complex indentation rules (which I <a href="https://github.com/weavejester/cljfmt">cljfmt</a> largely replicated).  I may end up doing something similar, but do wonder if some flexibility could be sacrificed in favour of easier customisation and better performance...
</p>

<p>
(<code>letfn</code> is actually even worse than the other <i>complex indentation</i> rules as its rules are unique!  In the old clojure.vim indentation code we had <a href="https://github.com/clojure-vim/clojure.vim/blob/bc3f1faf93ac47343b66b0e189990149a7fc0715/indent/clojure.vim#L175-L199">hardcoded rules</a> just to handle it!)
</p>
</details>

<details>
<summary>
Click this to view a complete tangent on my opinions about <code>letfn</code> overuse (that no one asked for).  🙂
</summary>
<br>
<p>
Personally I feel that <code>letfn</code> is overused in Clojure code, as the only reason I can see to use it over plain <code>let</code> + <code>fn</code> is when the <code>letfn</code>'d functions need to call each other.  If it were used only in that case, it could be useful as a signal that that is happening, much like how use of <code>when</code> over <code>if</code> signifies that there is no "else" branch.
</p>
<p>
(The equivalent of <code>letfn</code> in Scheme is <code>letrec</code> — i.e. let recursive — which I believe may have inspired <code>letfn</code> and kind of backs up my thoughts on its overuse.)
</p>
</details>

**Comment by garrett-hopper** [comment]
FYI, the `"uniform"` `indent_style` isn't quite following Tonsky's [Better Clojure formatting](https://tonsky.me/blog/clojurefmt/)

> Multi-line lists that start with a symbol are always indented with two spaces,
> Other multi-line lists, vectors, maps and sets are aligned with the first element (1 or 2 spaces).

Multi-line lists that start with a non-symbol are currently indented with 2 instead of 1 space.

E.g.

```clojure
(ns namespace
  (:require
   [clojure.set :as s]))
```

The final line is indented 1 space because `:require` is keyword and not a symbol.

```vim
if indent_style ==# 'uniform' | return base_indent + 1 | endif
```

Not sure the best way way for this to check if the first list item is a symbol or not. 🤔

I wrote this for #21 before I realized this rewrite was being done. It's a pattern that will match symbols, adapted from the tree-sitter grammar. (Though it may be easier to just special case non-symbols as first list item.)

```lua
local suffixonly = ":#'0-9\\/" -- not allowed as first character
local forbidden = {
  [[\[\]\n\r\t \\]], -- brackets, whitespace, slashes
  '(){}"@~^;`,', -- parens, braces, special characters
  suffixonly,
}
local prefix = string.format("[^%s]", table.concat(forbidden))
local suffix = string.format([[\(%s\|[%s]\)*]], prefix, suffixonly)
local symbol = prefix .. suffix
```

```
[^\[\]\n\r\t \\(){}"@~^;`,:#'0-9\/]\([^\[\]\n\r\t \\(){}"@~^;`,:#'0-9\/]\|[:#'0-9\/]\)*
```

**Comment by axvr** [comment]
Thanks for raising this @garrett-hopper.  I'll add it to my list and try to fix it soon.

(Yet more mildly irritating rules in the common Clojure indentation styles :laughing: :sob:)

**Comment by axvr** [comment]
Apologies to anyone waiting patiently for this to be merged.  I haven't managed to find much time to dedicate to this and each time I come back to it, it requires a lot of context reloading as the standard Clojure indentation styles are shockingly all very complicated (contrary to Clojure itself) and to implement them efficiently in Vim script is forever painful.

Currently what exists in this branch is IMO already far superior to what is currently on the `master` branch, so my current plan is:

1. fix a few of the more major outstanding issues,
2. document the `clojure_indent_rules` option with warnings that it may change in the future,
3. merge, 🥳 
4. open issues for the remaining known problems that I don't consider to be dealbreakers,
5. gather more feedback on other issues with wider use,
6. fix issues critical to the core algorithm,
7. later complete #35 and #32 once the core algorithm is perfected to avoid needing to implement the same fixes 3 times.

I hope to get round to the first 4 soon, but can't make any promises on when I will do so.

**Comment by NoahTheDuke** [comment]
Pretty exciting! Thanks for shepherding this! I've been using this branch exclusively since you opened the PR and I'm quite pleased with it.

**Comment by NoahTheDuke** [comment]
Bug report: defining protocol methods in `defrecord` under "standard" still indents the body to match the arg vector:

```clojure
(defrecord Example [a b c]
  MyProto
  (method1 [this d e f]
    (+ a b c d e f)))
```
becomes
```clojure
(defrecord Example [a b c]
  MyProto
  (method1 [this d e f]
           (+ a b c d e f)))
```

It looks like there's tests covering this, so I'm not sure what's wrong.
